source/lac/sparsity_pattern_base.cc

```diff
@@ -0,0 +1,79 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 - 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/lac/sparsity_pattern_base.h>
+
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <boost/container/small_vector.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
+
+#include <algorithm>
+#include <utility>
+
+DEAL_II_NAMESPACE_OPEN
+
+void
+SparsityPatternBase::add_entries(const ArrayView<const size_type> &rows,
+                                 const ArrayView<const size_type> &columns)
+{
+  AssertDimension(rows.size(), columns.size());
+  for (std::size_t i = 0; i < rows.size(); ++i)
+    add_row_entries(rows[i],
+                    make_array_view(columns.begin() + i,
+                                    columns.begin() + i + 1));
+}
+
+
+
+void
+SparsityPatternBase::sort_and_add_entries(
+  const ArrayView<std::pair<size_type, size_type>> &entries)
+{
+  // We always want to sort so that we can use the optimized add_row_entries()
+  // function
+  std::sort(entries.begin(), entries.end());
+  boost::container::small_vector<size_type, 128> columns;
+
+  auto entry = entries.begin();
+  while (entry != entries.end())
+    {
+      const auto row_end =
+        std::upper_bound(entry + 1,
+                         entries.end(),
+                         *entry,
+                         [](const std::pair<size_type, size_type> &a,
+                            const std::pair<size_type, size_type> &b) {
+                           return a.first < b.first;
```

Comment:
We don't generally add hundreds of entries. I suspect that you'd be faster if you just search linearly forward with a `for` loop that you `break` when you hit the next row.